src/main/java/org/spongepowered/api/item/inventory/Inventory.java

```diff
@@ -0,0 +1,263 @@
+ * 
+ * TODO Flesh out javadoc from proposal document. For now, see proposal doc
+ * here: https://github.com/SpongePowered/SpongeAPI/pull/443
+ */
+public interface Inventory extends Iterable<Inventory>, Nameable {
+    
+    /**
+     * Get the parent {@link Inventory} of this {@link Inventory} 
+     */
+    Inventory parent();
+
+    /**
+     * Returns an iterable view of all {@link Slot}s (leaf nodes) in this
+     * Inventory
+     */
+    <T extends Inventory> Iterable<T> slots();
+
+    /**
+     * Return the first child inventory, effectively the same as
+     * <code>Inventory::iterator().next()</code> but more convenient when we are
+     * expecting a result set with only a single entry. Also use type specifier
+     * to allow easy pseudo-duck-typing. If no children, then returns
+     * <code>this</code>.
+     */
+    <T extends Inventory> T first();
+
+    /**
+     * Return the next sibling inventory, allows traversing the inventory
+     * hierarchy without using an iterator. If no more children, returns an
+     * {@link EmptyInventory}.
+     */
+    <T extends Inventory> T next();
+
+    /**
+     * Get and remove the first available stack from this Inventory
+     */
+    Optional<ItemStack> poll();
+    
+    /**
+     * <p>Get and remove up to <code>limit</code> items of the type in the first
+     * available stack in this Inventory from all stacks in this Inventory. If
+     * no stack is available then {@link Optional#absent()} is returned (as per
+     * the usual behaviour of {@link #poll()}, otherwise a new {@link ItemStack}
+     * is returned containing the removed items, the contents of the stack in
+     * the inventory are reduced by the number of items consumed. Note that this
+     * method attempts to consume items into the ouput up to <code>limit</code>,
+     * which may consume items from an abitrary number of internal slots.</p>
+     * 
+     * <p>For example, assume an inventory containing 4 slots contains stacks as
+     * follows:</p>
+     * 
+     * <blockquote>
+     *     <pre>[Stone x10] [Dirt x3] [Arrows x9] [Stone x32]</pre>
+     * </blockquote>
+     * 
+     * <p>Calling <code>poll(16)</code> on this inventory will consume <em>Stone
+     * </em> from the Inventory (because the first stack contains stone), and
+     * will then consume the remaining 6 items from the 4th slot.</p>
+     * 
+     * <p>It is intended that this method is used in conjunction with a query
+     * which returns a set of slots containing a specific item type:</p>
+     * 
+     * <blockquote>
+     *     <pre>Optional&lt;ItemStack&gt; q = inv.query(ItemTypes.DIRT).poll(1);
+     *     </pre>
+     * </blockquote>
+     * 
+     * @param limit Maximum number of items to consume from the stack
+     */
+    Optional<ItemStack> poll(int limit);
+    
+    /**
+     * Get without removing the first available stack from this Inventory
+     */
+    Optional<ItemStack> peek();
```

Comment:
You should mention whether this returns a copy or a life item. Otherwise the following code's result is undefined.
````java
inventory.query(ItemTypes.ARROW).peek().setQuantity(64);
````